optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/common/UniLeftHandSide.java

```diff
@@ -121,51 +121,37 @@ protected UniLeftHandSide(PatternVariable<A> patternVariable, DroolsVariableFact
         return new UniLeftHandSide<>(this, patternVariable.filter(predicate));
     }
 
-    private <B> UniLeftHandSide<A> applyJoiners(Class<B> bClass, AbstractBiJoiner<A, B> joiner,
+    private <B> UniLeftHandSide<A> applyJoiners(Class<B> otherFactType, AbstractBiJoiner<A, B> joiner,
             BiPredicate<A, B> predicate, boolean shouldExist) {
-        Variable<B> toExist = (Variable<B>) variableFactory.createVariable(bClass, "toExist");
+        Variable<B> toExist = (Variable<B>) variableFactory.createVariable(otherFactType, "toExist");
         PatternDSL.PatternDef<B> existencePattern = pattern(toExist);
         if (joiner == null) {
-            return applyFilters(patternVariable, existencePattern, predicate, shouldExist);
+            return applyFilters(existencePattern, predicate, shouldExist);
         }
-        PatternVariable<A> newPatternVariable = patternVariable;
         JoinerType[] joinerTypes = joiner.getJoinerTypes();
-        // Rebuild the A pattern, binding variables for left parts of the joins.
-        Variable[] joinVars = new Variable[joinerTypes.length];
         for (int mappingIndex = 0; mappingIndex < joinerTypes.length; mappingIndex++) {
-            // For each mapping, bind one join variable.
-            Variable<Object> joinVar = variableFactory.createVariable("joinVar");
-            Function<A, Object> leftMapping = joiner.getLeftMapping(mappingIndex);
-            newPatternVariable = newPatternVariable.bind(joinVar, leftMapping);
-            joinVars[mappingIndex] = joinVar;
-        }
-        // Create the B pattern, joining with the new A pattern using its freshly bound join variables.
-        for (int mappingIndex = 0; mappingIndex < joinerTypes.length; mappingIndex++) {
-            // For each mapping, bind a join variable from A to B and index the binding.
```

Comment:
How was such a simplification possible?

Reply:
Good question. The original code was a leftover from a much more complicated world of the ancient CS-D implementation.
In this simplified CS-D, many things are different and one of them is that I can use variables directly instead of proxying them.